runtime/protocol/http-client-engines/test-suite/jvm/src/aws/smithy/kotlin/runtime/http/test/util/AbstractEngineTestJVM.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.http.test.util
+
+import aws.smithy.kotlin.runtime.http.engine.HttpClientEngine
+import aws.smithy.kotlin.runtime.http.engine.ktor.KtorEngine
+import aws.smithy.kotlin.runtime.http.sdkHttpClient
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeout
+import kotlin.coroutines.CoroutineContext
+import kotlin.time.Duration
+
+actual abstract class AbstractEngineTest actual constructor() {
+    actual fun testEngines(block: EngineTestBuilder.() -> Unit) {
+        engines().forEach { engine ->
+            val client = sdkHttpClient(engine)
+            testWithClient(client, block = block)
+        }
+    }
```

Comment:
**Question**: This implementation like it would be the same for any platform...iterate through the list of available engines, construct a client, and run the test. What other actions could a platform want to take? Would it be better to make `engines()` the `expect`/`actual` method and pull `testEngines` up into common code?

Reply:
If you look at the ktor equivalent they have an additional parameter to allow ignoring specific engines for a test (they also have specific debugging per/target). If we don't think we'll ever need per/platform behavior in here then we could move it up

Reply:
Seems like ignoring engines for some tests could also be handled by making `engines()` non-final.

Reply:
Well the other reason would be to allow per/test engine configuration. Right now we just statically create `KtorEngine` with a default `HttpClientEngineConfig`. Which means we'll need to change to a factory pattern for `engines(config: HttpClientConfig)`.

I'll pull it up for now though since I don't see us needing to skip engines yet (since the whole point here is to assert expected behaviors across all supported engines)